gnovm/stdlibs/std/emit_event.go

```diff
@@ -0,0 +1,75 @@
+package std
+
+// ref: https://github.com/gnolang/gno/pull/575
+// ref: https://github.com/gnolang/gno/pull/1833
+
+import (
+	"encoding/json"
+	"errors"
+
+	gno "github.com/gnolang/gno/gnovm/pkg/gnolang"
+	abci "github.com/gnolang/gno/tm2/pkg/bft/abci/types"
+)
+
+func X_emit(m *gno.Machine, typ string, attrs []string) {
+	eventAttrs, err := attrKeysAndValues(attrs)
+	if err != nil {
+		m.Panic(typedString(err.Error()))
+	}
+
+	pkgPath := CurrentRealmPath(m)
+	fnIdent := getPrevFunctionNameFromTarget(m, "Emit")
+
+	evt := NewGnoEvent(typ, pkgPath, fnIdent, eventAttrs...)
+	ctx := m.Context.(ExecContext)
+	ctx.EventLogger.EmitEvent(evt)
+
+	str, err := json.Marshal(evt)
+	if err != nil {
+		m.Panic(typedString(err.Error()))
+	}
+
+	ctx.EventString = abci.EventString(str)
+}
+
+func attrKeysAndValues(attrs []string) ([]gnoEventAttribute, error) {
+	attrLen := len(attrs)
+	if attrLen%2 != 0 {
+		return nil, errors.New("cannot pair attributes due to odd count")
+	}
+	eventAttrs := make([]gnoEventAttribute, attrLen/2)
+	for i := 0; i < attrLen-1; i += 2 {
+		eventAttrs[i/2] = gnoEventAttribute{
+			Key:   attrs[i],
+			Value: attrs[i+1],
+		}
+	}
+	return eventAttrs, nil
+}
+
+func NewGnoEvent(eventType, pkgPath, ident string, attrs ...gnoEventAttribute) *gnoEvent {
+	return newGnoEvent(eventType, pkgPath, ident, attrs...)
+}
```

Comment:
Remove.

If you need to use this outside of the package, export `newGnoEvent`, `gnoEvent` and `gnoEventAttribute` as well.

Reply:
Removed the capital New function and modified it to use `newGnoEvent` function directly.